pkg/engine/datasource/graphql_datasource/graphql_datasource.go

```diff
@@ -1342,31 +1343,33 @@ func (s *Source) compactAndUnNullVariables(input []byte) []byte {
 	}
 
 	removeNullVariables := httpclient.IsInputFlagSet(input, httpclient.UNNULLVARIABLES)
-	variables = s.cleanupVariables(variables, removeNullVariables)
+	variables = s.cleanupVariables(variables, removeNullVariables, undefinedVariables)
 
 	input, _ = jsonparser.Set(input, variables, "body", "variables")
 	return input
 }
 
 // cleanupVariables removes null variables and empty objects from the input if removeNullVariables is true
 // otherwise returns the input as is
-func (s *Source) cleanupVariables(variables []byte, removeNullVariables bool) []byte {
+func (s *Source) cleanupVariables(variables []byte, removeNullVariables bool, undefinedVariables []string) []byte {
 	cp := make([]byte, len(variables))
 	copy(cp, variables)
 
-	if removeNullVariables {
-		// remove null variables from JSON: {"a":null,"b":1} -> {"b":1}
-		err := jsonparser.ObjectEach(variables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
-			if dataType == jsonparser.Null {
+	// remove null variables from JSON: {"a":null,"b":1} -> {"b":1}
+	err := jsonparser.ObjectEach(variables, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
+		if dataType == jsonparser.Null {
+			if removeNullVariables || slices.Contains(undefinedVariables, string(key)) {
```

Comment:
as linter fails with generics we could use instead smth like this

```
type UndefinedVars []string

func(vars UndefinedVars) Contains(name string) (bool) {
  for i := 0; i < len(vars); i++ {
    if vars[i] == name {
      return true
    }
  }
  return false
}

Reply:
fixed by updating CI